### ISSUE TITLE
Requested watcher progress notifications are not synchronised with stream

### ISSUE THREAD

**Issue by scpmw** [etcd-io/etcd#15220]
### What happened?

When you request a progress notification from a watch stream, the header of the generated progress notification can indicate a revision for which not all watch responses have been generated yet.

This means that there is a (rare) race here - if the client receives the progress notification, and the connection drops immediately, it will presumably assume that it has received all updates in the database up to the revision indicated in the progress notification. It might therefore create a new watcher starting from this revision, which will cause a watch notification to be dropped.

A more easily reproduced variant of this problem actually happens when you create a new watcher with a start revision and immediately request a progress notification: The progress notification will typically arrive **before** any watch responses between the start revision and the current revision.

It looks like this might be because the implementation of the response code in https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L363 immediately sends the response through `sws.ctrlStream`, which can overtake `sws.watchStream` for the purpose of the `sendLoop`.

### What did you expect to happen?

Note that this only happens for manually requested progress notifications. Timed progress notifications seem to go to great pain to ensure that the watchers are synced: The `progressTicker` in `sendLoop` uses `sws.watchStream.RequestProgress(id)`, which makes sure the watcher is synched and that the progress notification is ordered correctly relative to other watch responses.

In fact, it seems like the immediate problem can be fixed by changing how `pb.WatchRequest_ProgressRequest` is handled?
```go
			sws.mu.Lock()
			for id, _ := range sws.watchers {
                                sws.watchStream.RequestProgress(id)
 			}
			sws.mu.Unlock()
```

Where `sws.watchers` is a map introduced just to keep track of *all* active watchers. Not sure this is actually the right fix - at minimum it changes the semantics of progress notifications in some more-or-less subtle ways:
* Now we get separate progress notifications for every watcher (ID)
* Because `progress` in `watchable_store.go` just skips the progress notification for unsynced clients, it is no longer guaranteed that every progress notification request actually generates a progress notification

Not sure whether this would be a problem?

### How can we reproduce it (as minimally and precisely as possible)?

Here is a Python script that reproduces the issue:
```python
import queue
import rpc_pb2_grpc
import rpc_pb2
import grpc
# Open channel, create watch
with grpc.insecure_channel('localhost:12346') as channel:
    stub = rpc_pb2_grpc.WatchStub(channel)
    kv = rpc_pb2_grpc.KVStub(channel)
    test_key = b"test"
    out_queue = queue.Queue()
    out_queue.put(
        rpc_pb2.WatchRequest(create_request=rpc_pb2.WatchCreateRequest(
            key=test_key,
            range_end=test_key + b"\ff"
    )))
    # Receive watch events
    last_rev = None
    for i, resp in enumerate(stub.Watch(iter(out_queue.get, None))):
        # Log message
        print(resp)
        # Check for out-of-order
        if last_rev is not None and resp.header.revision < last_rev and resp.events:
            print(" ^^^ Out-of-order watch response found")
            exit(1)
        # Change key & request progress notificiation
        if last_rev is None or last_rev != resp.header.revision:
            out_queue.put(
                rpc_pb2.WatchRequest(progress_request=rpc_pb2.WatchProgressRequest()))
            kv.Put(rpc_pb2.PutRequest(key=test_key,value=str(i).encode()))
            last_rev = resp.header.revision

```

You might need to run multiple instances for a while to put enough stress on the server, but eventually you will get something like this:
```
header {
  cluster_id: 14841639068965178418
  member_id: 10276657743932975437
  revision: 373837
  raft_term: 2
}
watch_id: -1

header {
  cluster_id: 14841639068965178418
  member_id: 10276657743932975437
  revision: 373835
  raft_term: 2
}
events {
  kv {
    key: "test"
    create_revision: 2
    mod_revision: 373835
    version: 373834
    value: "46952"
  }
}

 ^^^ Out-of-order watch response found
```
Which demonstrates that the problem can happen: We have a progress notification for revision 373837 *before* a genuine watch notification for revision 373835.

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
etcd Version: 3.5.0-pre
Git SHA: 772dfbfe3
Go Version: go1.18.1
Go OS/Arch: linux/amd64

$ etcdctl version
etcdctl version: 3.4.5
API version: 3.4
```
</details>

**Comment by serathius** [comment]
QQ Why you are using etcd server version `3.5.0-pre`? There have been many data inconsistencies found in v3.5.X releases. Please repeat the test on latest releases v3.4.23 v3.5.7

**Comment by serathius** [comment]
> You might need to run multiple instances for a while to put enough stress on the server.

Can you give some idea of how much qps is `enough stress`?

**Comment by scpmw** [comment]
I tested both 3.3.25, then used a master version to test whether my suggestion worked. 3 instances of the reproduction script running in parallel end up doing around 200 revisions per second. Can take a couple minutes for the race to occur.

**Comment by serathius** [comment]
What do you mean by by `master version`?

> 3 instances of the reproduction script running in parallel end up doing around 200 revisions per second.

That's not a lot. Existing linearizability tests make 1k qps/s and validate that watch response between members match. We don't directly validate that revision is strictly increasing, so I can work on adding it.

**Comment by scpmw** [comment]
As in, the current `main` branch. Apologies, had not checked what it was called here.

And yes, most of the code is taking quite a bit of care to guarantee ordering, which is why I felt it was so out of place that requested progress notifications just short-cut all of that and directly send out an answer with the "current" revision in the header. For a `CreateRequest` or `CancelRequest` you presumably don't really care what the `revision` in the `header` is, but for a progress notification that revision is the entire point, so it seems like some more care should be taken.

**Comment by scpmw** [comment]
Here is a slightly more serious attempt at a fix: https://github.com/scpmw/etcd/commit/cd3b22ff79cc37597890c6e13429a1196ef72176

This mostly retains the old behaviour - you get responses with WatchId -1, and unsynced clients would no longer just ignore progress notification requests. Main difference is that it is no longer guaranteed that a given number of progress notification requests will be matched by an equal number of progress notifications: If the watcher happens to be unsynced, we just set a flag for a notification to be sent later, which means that multiple requests might get rolled into one.

**Comment by serathius** [comment]
cc @logicalhan

**Comment by serathius** [comment]
cc @wojtek-t @lavalamp @hankang

**Comment by serathius** [comment]
Looking into reproducing it myself.

**Comment by serathius** [comment]
It immediately reproduces for me. Just sending a RequestProgress results in watch getting a response with lower revision. Important to note, there are to ProgressNotification features in etcd:
* [progress_notify flag on WatchCreateRequest](https://github.com/etcd-io/etcd/blob/f25b159fe3e95e07ea9b49cde765bbee928b24e7/api/etcdserverpb/rpc.proto#L759-L763) that ensures that server sends at least one response every 10 minutes.
* [WatchProgressRequest](https://github.com/etcd-io/etcd/blob/main/api/etcdserverpb/rpc.proto#L800) allows client to immediately request a notification.

Assessing impact. I think Kubernetes uses only the progress_notify flag and never the second one `RequestProgress`. Looking into confirming which etcd versions it affects.

**Comment by serathius** [comment]
cc @ptabor @ahrtr

**Comment by serathius** [comment]
@scpmw Feel free to open an PR with the fix you proposed. We can discuss the details there. Would be great to also add an e2e tests that confirms the fix

**Comment by serathius** [comment]
cc @mborsz

**Comment by ahrtr** [comment]
Thanks @scpmw for raising this issue. 

Firstly, this is a real issue to me, but not a critical one to `K8s`, probably not even a major issue. Reasons:
1. K8s doesn't use [(*watcher) RequestProgress](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/client/v3/watch.go#L410) at all. 
2. K8s indeed uses [clientv3.WithProgressNotify()](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/client/v3/op.go#L477), but etcdserver will skip the progress notification if a normal WatchResponse (with at least one event) has been sent after last progress notification being sent.
3. I agree there is indeed a race condition that a progress notification with a higher rev might be sent, and followed with a WatchResponse with a lower rev. But the default progress notification interval is 10+ minutes (configured by `--experimental-watch-progress-notify-interval` (what's the value set by K8s?), and defaults to 10min + [jitter](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L99) ), see [config.go#L275](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdmain/config.go#L275) and [watch.go#L86](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L86), so it should be hard (at least not frequently) to reproduce this issue in production. 
4. Take a step back, even it happens, the client side can just ignore it. Note the progress notification will always have 0 [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217), while a normal WatchResponse will always have 1+ [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217). I am not sure how K8s process each progress notification response, which is actually a WatchResponse which has 0 event. Can anyone double confirm this?

The solution provided in https://github.com/scpmw/etcd/commit/guarantee_progress_notify_order overall looks good to me, but with a couple of comments. Please feel free to deliver a PR @scpmw . Please add an E2E test to reproduce this issue firstly, and then apply your fix to confirm that it indeed fixes the issue.

cc  @dims @liggitt @nikhita @neolit123

**Comment by ahrtr** [comment]
cc @mitake

**Comment by wojtek-t** [comment]
Thanks for bringing this up to my attention.

I think that this part is critical for me:

> Note that this only happens for manually requested progress notifications. Timed progress notifications seem to go to great pain to ensure that the watchers are synced: 

In Kubernetes we're only using the timed progress notifications - we're not using manually requests ones, so this doesn't seem to impact Kubernetes itself.

@serathius @lavalamp ^^

**Comment by scpmw** [comment]
Yes, this is only for manually requested progress notifications, which I understand is a relatively new and niche feature.

I'll check whether I can come up with a good integration test - might well focus on the specific case of requesting a progress notification right after creating a watcher, as that is the easiest case to reproduce.

**Comment by ahrtr** [comment]
> Yes, this is only for manually requested progress notifications

This isn't true. The time based progress notification also has this issue, but just as I mentioned [above](https://github.com/etcd-io/etcd/issues/15220#issuecomment-1412894572), it's hard to reproduce this issue in production.

**Comment by scpmw** [comment]
It does? In which case? It was my impression that because
1. timed progress notifications go through the same `watchStream` as all other watch responses and
2. are held back if the stream isn't synched (see https://github.com/scpmw/etcd/blob/8de21779ba755c251f2cd5a4383ca99d31aeb7c0/mvcc/watchable_store.go#L495 )

order should be correct. Have I overlooked something? Because if so, my fix would also be wrong.

**Comment by wojtek-t** [comment]
ohh sorry - I missed the last few comments, seems I opened the window when they weren't there and didn't refresh

> But the default progress notification interval is 10+ minutes (configured by --experimental-watch-progress-notify-interval (what's the value set by K8s?), and defaults to 10min + [jitter](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L99) ), see [config.go#L275](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdmain/config.go#L275) and [watch.go#L86](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L86), so it should be hard (at least not frequently) to reproduce this issue in production.

In k8s we're setting it to 5s, but 1s is something that we would like to use (and it seemed to work in our tests).


> Take a step back, even it happens, the client side can just ignore it. Note the progress notification will always have 0 [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217), while a normal WatchResponse will always have 1+ [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217). I am not sure how K8s process each progress notification response, which is actually a WatchResponse which has 0 event. Can anyone double confirm this?

What do you mean by "client side can just ignore it"?
The whole purpose of enabling this feature is to actually get those notifications, not ignore them...

They way we're using this feature is to keep up with RV changes - we're watching for a single resource type (a subset of objects stored in etcd) and we want to know to what RV we're actually synced to if the set of objects is changing very infrequently.
So basically for sets of objects don't change, we want to ensure that the RV to which we know we're synced is not too far in the past (e.g. to make restoration on failure cheaper and to be able to serve data from cache).

If it's possible that the progress-notify can come with a higher RV than the watch event that will come later, that is a huge correctness issue (because if watch breaks in the meantime, we will resume from the last RV we received so far - the assumption is that RVs we receive are monotonically growing).

**Comment by serathius** [comment]
> ohh sorry - I missed the last few comments, seems I opened the window when they weren't there and didn't refresh
> 
> > But the default progress notification interval is 10+ minutes (configured by --experimental-watch-progress-notify-interval (what's the value set by K8s?), and defaults to 10min + [jitter](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L99) ), see [config.go#L275](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdmain/config.go#L275) and [watch.go#L86](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/server/etcdserver/api/v3rpc/watch.go#L86), so it should be hard (at least not frequently) to reproduce this issue in production.
> 
> In k8s we're setting it to 5s, but 1s is something that we would like to use (and it seemed to work in our tests).

Based on my experiments and the original comment I **don't think the bug affects timed progress notification at all**.

**Comment by serathius** [comment]
Repros:
* https://github.com/etcd-io/etcd/pull/15235
* https://github.com/etcd-io/etcd/pull/15234

**Comment by ahrtr** [comment]
Let's work with an example workflow.

1. A client just puts some K/V pairs; and the rev already increased;
2. etcdserver needs to notify all watchers, see [watchable_store_txn.go#L43-L44](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store_txn.go#L43-L44). 
3. Assuming at the moment when step2 just is in a state between line 43 and 44, the [timed progress notification ticker is triggered](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/etcdserver/api/v3rpc/watch.go#L514), accordingly it gets the latest `rev`. Note that if the watcher streams number <= [128](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store.go#L35), [it sends the events directly and doesn't remove any watcher from synced group at all](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store.go#L447-L456).
4. Assuming step 2 continue to execute line 44, then the watcher might get the same revision. 

Unfortunately, I do not get time to reproduce this issue. 

Obviously it's really hard to reproduce this issue. Suggestions to reproduce this issue:
1. Create lots of client/watch streams, make sure > 128;
2. Lower the progress notification interval;
3. Create large volume of traffic

**Comment by ahrtr** [comment]
> What do you mean by "client side can just ignore it"?

I was saying to ignore the error when you see an rev <= one of the historical rev if the latter (the historical value) is coming from a progress notification response. 

Just as I mentioned in https://github.com/etcd-io/etcd/issues/15220#issuecomment-1412894572. The progress notification will always have 0 [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217), while a normal WatchResponse will always have 1+ [events](https://github.com/etcd-io/etcd/blob/b10ebff3bb0c337b3b153a39474e33fa9698790a/api/etcdserverpb/rpc.pb.go#L2217)

Based on my above comment https://github.com/etcd-io/etcd/issues/15220#issuecomment-1413563670, it seems that it isn't possible to get a lower rev, but it's possible for the client to get the same rev twice. Again, it's still hard to reproduce it in production, even in development env.

**Comment by serathius** [comment]
> Based on my above comment https://github.com/etcd-io/etcd/issues/15220#issuecomment-1413563670, it seems that it isn't possible to get a lower rev for timed progress notification, but it's possible for the client to get the same rev twice

That doesn't seem problematic at all.

**Comment by wojtek-t** [comment]
> That doesn't seem problematic at all. Revision doesn't need to be strictly increasing, it's enough that it's non-decreasing.

Not really, if the first one is progress notification, and the latter is watch event
[as then if watch breaks in the middle, you will resume from the next one and miss that watch event.]

**Comment by ahrtr** [comment]
Let's get K8s folks feedback on this

> That doesn't seem problematic at all. Revision doesn't need to be strictly increasing, it's enough that it's non-decreasing.

**Comment by serathius** [comment]
> > That doesn't seem problematic at all. Revision doesn't need to be strictly increasing, it's enough that it's non-decreasing.
> 
> Not really, if the first one is progress notification, and the latter is watch event [as then if watch breaks in the middle, you will resume from the next one and miss that watch event.]

Makes sense. I will try to reproduce this.

**Comment by serathius** [comment]
Pushed https://github.com/etcd-io/etcd/pull/15234 with validating order of watch event and progress notify

I was able to reproduce case with same revision coming twice without any issue. Fortunately for now it looks that progress notify only comes as the second response.

**Comment by wojtek-t** [comment]
> now it looks that progress notify only comes as the second response.

This is fine.

**Comment by serathius** [comment]
To summarize:
* We have confirmed that manually requested progress notification can return lower revision than preceding watch event. Reproduced in https://github.com/etcd-io/etcd/pull/15235
* We also confirmed that enabling timed progress notify can result in same revision being returned once. Should not be surprising, if no changes happened since last watch event then progress notify will repeat last revision.
* What's left to confirm, is whether there can be a race that allows progress notify be returned before watch event with the same revision.

I don't have good knowledge of this part of codebase, but from what I see there should be no case for timed progress notify to read revision from watch event before watch event is sent.

Reason:
* [Watchers are notified before revision is updated in storage](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store_txn.go#L44-L45)
* [Sending watch response (both events and progress) is done via a single channel](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store.go#L537-L545) and if it doesn't succeed [watcher is removed from synced watcher list](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store.go#L447-L456) 
* [Progress is only sent to synced watchers](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store.go#L474-L483)

**Comment by serathius** [comment]
Cannot test https://github.com/scpmw/etcd/commit/guarantee_progress_notify_order because it hangs on old etcd version v3.4.10. @scpmw please cherry-pick it on main branch as I have confirmed that it's affected.

**Comment by scpmw** [comment]
Oh, apologies. Fixed - unfortunately removing some existing discussion. I've also opened a PR for the proposal: https://github.com/etcd-io/etcd/pull/15237

This won't do anything for the issue @ahrtr raises, that might need an additional fix.

**Comment by ahrtr** [comment]
> [Watchers are notified before revision is updated in storage](https://github.com/etcd-io/etcd/blob/9505d729302467b393a74190741d57c5fd5db639/server/storage/mvcc/watchable_store_txn.go#L44-L45)

This is the key point. Previously I misread it that revision is updated before watchers are notified. 

**Confirmed that timed progress notification has no issue.**

**Comment by logicalhan** [comment]
> Revision doesn't need to be strictly increasing, it's enough that it's non-decreasing.

The way k8s uses it, it must be strictly increasing.

**Comment by logicalhan** [comment]
> We also confirmed that enabling timed progress notify can result in same revision being returned once. Should not be surprising, if no changes happened since last watch event then progress notify will repeat last revision.

Do you mean `returned twice`?

**Comment by wojtek-t** [comment]
> Confirmed that timed progress notification has no issue.

Thanks - that means that from k8s perspective we're good then.

**Comment by ahrtr** [comment]
> Thanks - that means that from k8s perspective we're good then.

Exactly.

**Comment by serathius** [comment]
> > We also confirmed that enabling timed progress notify can result in same revision being returned once. Should not be surprising, if no changes happened since last watch event then progress notify will repeat last revision.
> 
> Do you mean `returned twice`?

@logicalhan, K8s by default configures timed watch progress notification used to get revision on resources that are not frequently changed. If there was no events on watched resource, etcd will send an empty watch response that will include revision. If there were no changes in etcd, revision might match revision from previous event.

**Comment by logicalhan** [comment]
> > > We also confirmed that enabling timed progress notify can result in same revision being returned once. Should not be surprising, if no changes happened since last watch event then progress notify will repeat last revision.
> > 
> > 
> > Do you mean `returned twice`?
> 
> @logicalhan, K8s by default configures timed watch progress notification used to get revision on resources that are not frequently changed. If there was no events on watched resource, etcd will send an empty watch response that will include revision. If there were no changes in etcd, revision might match revision from previous event. So from client perspective, watch response might include same revision multiple times with the first one including watch event, and the other ones being progress notification.

As long as a bookmark RV doesn't match a **subsequent** event, then that should be okay.
